src/nvim/main.c

```diff
@@ -355,11 +349,19 @@ int main(int argc, char **argv)
     ui_client_channel_id = rv;
   }
 
+  // NORETURN: Start builtin UI client.
   if (ui_client_channel_id) {
     time_finish();
     ui_client_run(remote_ui);  // NORETURN
   }
   assert(!ui_client_channel_id && !use_builtin_ui);
+  // Nvim server...
+
+  int listen_rv = server_init(params.listen_addr);
+  if (listen_rv != 0) {
+    mainerr("Failed to --listen", listen_rv < 0
```

Comment:
Without `--headless`, this error message doesn't print nicely. But that same problem affects `nvim -W` validation and probably other startup validation.

TODO: if the server fails, can the UI client properly teardown the screen and print the server error?